openwisp_users/apps.py

```diff
@@ -109,22 +113,50 @@ def connect_receivers(self):
                         name, model.__name__
                     ),
                 )
+            pre_save.connect(
+                self.pre_save_update_organizations_dict,
+                sender=model,
+                dispatch_uid='{}_{}_pre_save_organizations_dict'.format(
+                    name, model.__name__
+                ),
+            )
         post_save.connect(
             self.create_organization_owner,
             sender=OrganizationUser,
             dispatch_uid='make_first_org_user_org_owner',
         )
 
     @classmethod
-    def update_organizations_dict(cls, instance, **kwargs):
+    def pre_save_update_organizations_dict(cls, instance, **kwargs):
+        """
+        Invalidates user's organizations cache, if
+        OrganizationUser.user or OrganizationOwner.organization_user
+        field are changed.
+        """
+
+        def _invalidate_old_related_obj_cache(instance, check_field):
+            Model = instance._meta.model
+            try:
+                db_obj = Model.objects.select_related(check_field).get(pk=instance.pk)
+            except Model.DoesNotExist:
+                return
+            else:
+                if getattr(db_obj, check_field) != getattr(instance, check_field):
+                    cls._invalidate_user_cache(getattr(db_obj, check_field))
+
         if hasattr(instance, 'user'):
-            user = instance.user
+            _invalidate_old_related_obj_cache(instance, 'user')
         else:
-            user = instance.organization_user.user
+            _invalidate_old_related_obj_cache(instance, 'organization_user')
+
+    @classmethod
+    def _invalidate_user_cache(cls, user):
+        User = get_user_model()
+        if not isinstance(user, User):
+            user = user.user
+
         cache_key = 'user_{}_organizations'.format(user.pk)
```

Comment:
The user's organizations dict is dependent on user's primary key. When we change the user field of an OrganizationUser, the cache for both the previous user and the new user have to be invalidated. Consider the following scenario:

An OrganizationUser object is created with `OrgA` and `User1` objects. The organization dict of `User1` is updated in the cache. Now, the same `OrganizationUser` object is changed and `User2` is used instead.  The old code only invalidated the organization dict cache for `User2`. Therefore, `User1` organization dict would reflect that it is sill a member of `OrgA`.